File: charper1 simple simulation/A1042 shuffling machine.py
k = int(raw_input())
N = 54
char = ['S','H','C','D','J']
start = range(1,N+1) #快速生成递增的一个列表
start = [0] + start
end = [0] * (N + 1)#快速生成相同元素
next = [0] * N
next = map(int, raw_input().split())
next = [0] + next

for i in range(k):
	for j in range(1, N + 1):
		end[next[j]] = start[j] #难点所在
	start = end
for i in range(N):
	print "%c%d" % (char[start[i+1] / 13], start[i+1] % 13 + 1),

'''第一次循环正确，但是为什么一进入第二次循环就乱了？明明在循环结束输出了一下
start是正确的，然后进入第二次循环，输出start[j]就错了？？？？'''

File: chapter1 simple simulation/A1042 shuffling machine.cpp
#include <cstdio>
const int N = 54;
int k;
int start[N], next[N], end[N];
char map[5] = {'S', 'H', 'C', 'D', 'J'};

int main(){
	scanf("%d", &k);
	for(int i = 1; i <= N; ++i) {
		start[i] = i;
	}
	for (int i = 1; i <= N; ++i){
		scanf("%d", &next[i]);
	}
	for (int i = 0; i < k; ++i){
		for (int j = 1; j <= N; ++j){
			end[next[j]] = start[j];
		}
		for(int j = 1; j <= N; ++j) {
			start[j] = end[j];
		}
	}
	for (int i = 1; i <= N; ++i){
		if(i != 1) printf(" ");
		start[i]--;
		printf("%c%d", map[start[i] / 13], start[i] % 13 + 1);
	}
	return 0;
	// for (int i = 1; i < N; ++i)
	// {
	// 	start[i]--;
	// 	printf("%c%d ", map[start[i] /13], start[i] % 13 + 1);
	// }
	// start[N]--;
	// printf("%c%d", map[start[N] /13], start[N] % 13 + 1);
	// return 0;
	//该格式和书中答案测试也是只有一个case通过。上面代码基本上和书中一模一样，还是不知道哪里错了
}
